File: app/graphql/mutations/add_to_cart.rb
module Mutations
  class AddToCart < BaseMutation
    argument :service_id, ID, required: true
    argument :appointment_id, ID, required: true
    argument :addon, String, required: false
    argument :slots, Integer, required: true

    type Types::Cart::AddToCartResultType, null: false

    def resolve(args = {})
      return :unauthenticated unless (user = context[:current_user])

      cart_item = user.cart.cart_items.build(**args)
      if cart_item.save
        cart_item
      else
        cart_item.errors
      end
    end
  end
end

File: app/javascript/graphql-requests/carts/add-to-cart.js
import apolloClient from "../../apollo";
import { ADD_TO_CART } from "../../graphql/mutations";

export default function addToCart({ serviceId, appointmentId, addon, slots }) {
  return apolloClient.mutate({
    mutation: ADD_TO_CART,
    variables: {
      input: {
        serviceId,
        appointmentId,
        addon,
        slots,
      },
    },
  });
}
